### ISSUE TITLE
Event and filtering system (high-level API)

### ISSUE THREAD

**Issue by graup** [aergoio/aergo#15]
We need an event system, not just for plugins like in #14, but also as general API so that clients can react without needing to scan all transactions.

### In Go

How about making a system similar to the actor messages with pub/sub? Maybe a component could
    `Subscribe(events.Block)`
and then
```go
func (ns *SomeComponent) Receive(context actor.Context) {
    switch msg := context.Message().(type) {
        case *events.Block:
            ns.HandleBlock(msg.Block);
        default:
    }
}
```

For contract events, you probably want to subscribe to something like Ethereum-style topics.
    `SubscribeWithFilter(events.Transaction, { To: someAddress, Topic: someTopic } )`

This is how it could look in Go, but more importantly this should also be accessible over GRPC, ideally supporting both streams for real-time and searching in the past.

### End-to-end example for streams

Herajs

    contract.events.Deposit({from: someAddress}).on('transaction', (tx) => { console.log(tx); });

GRPC

    wss://127.0.0.1:7845/types.AergoRPCService/EventsStream
    event=...   from=someAddress

Aergosvr

    SubscribeWithFilter(events.Transaction, { To: someAddress, Topic: someTopic } )
    Receive: 
        stream.Send(transaction)

(Need to think about performance here, what if there are a lot of clients connecting to the server at once?)

### Filtering

Is there any problem with just implementing Ethereum-like filters? (Store logs in tx receipts and have bloom filters to quickly search for topics)

Some related discussion about topics/filters:
- https://ethereum.stackexchange.com/questions/12950/what-are-event-topics
- https://github.com/ethereum/go-ethereum/blob/master/eth/filters/filter.go
- https://github.com/ethereum/go-ethereum/blob/86a03f97d30c11a5321fa2f0fd37cbc4c7f63a32/core/types/bloom9.go
- https://github.com/ethereum/go-ethereum/issues/17613

**Comment by ashen1dev** [comment]
We already have some plans to create an event framework. First one is Pub-Sub model like Ethereum. and Second one is POST method in Coinstack 3sp1. We can discuss them in here.

**Comment by graup** [comment]
@ashen1dev I think the POST method in Coinstack is interesting, but the use case seems a bit different. I guess it's a bit like a webhook in GitHub? So the server will request a given URL when something happens in the contract? It makes the contract very tightly coupled to a specific centralized deployment, right?

**Comment by graup** [comment]
This has landed in 0.12, but documentation is still lacking.

**Comment by graup** [comment]
There's some documentation here:

- https://docs.aergo.io/en/latest/specs/contracts.html
- https://docs.aergo.io/en/latest/smart-contracts/lua/reference.html#event-eventname-args

I'd like to see a bit more in the first article and also in the Lua Guide
